[HUDI-759] Integrate checkpoint provider with delta streamer

garyli1019 opened the pull request:
## What is the purpose of the pull request

Integrate the initial checkpoint provider with delta streamer

## Brief change log

  - Add two options to delta streamer to use the initial checkpoint provider

## Verify this pull request

This change added tests and can be verified as follows:

  - Added unit test in TestHoodieDeltaStreamer

## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

garyli1019 commented:
@pratyakshsharma @vinothchandar 
I agree with Vinoth's idea that having those two options in the delta streamer. Is this implementation makes sense to you guys?
I will add a test case if this approach looks good. 
Thanks

pratyakshsharma commented:
LGTM

garyli1019 commented:
Test added. Thanks for the review

garyli1019 commented:
Add https://github.com/apache/incubator-hudi/pull/1493 into this PR.

pratyakshsharma commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/deltastreamer/DeltaSync.java:
```diff
@@ -153,7 +153,7 @@
 
   public DeltaSync(HoodieDeltaStreamer.Config cfg, SparkSession sparkSession, SchemaProvider schemaProvider,
-                   TypedProperties props, JavaSparkContext jssc, FileSystem fs, HiveConf hiveConf,
+                   TypedProperties props, JavaSparkContext jssc, FileSystem fs, Configuration hiveConf,
```
Let us change the variable name to conf?

garyli1019 replied:
I keep `hiveConf` here for better readability. There are two other configuration `cfg` `props` here, so I believe `hiveConf` could be more clear for the user to know what this is for. WDYT?

vinothchandar replied:
but it can be misleading for e.g if user does not even configure hive sync right? I also prefer `conf`

garyli1019 replied:
This config was only used by hive sync, if the user don't config this it will use the default Hadoop config. I don't have a strong opinion here and ok to go with `conf`.

garyli1019 replied:
ok...After a closer look, I'd still vote for naming this as `hiveConf`. There are `cfg` `props` in `HoodieDeltaStreamer` and `DeltaSync`. If we name it conf, I personally will get confused while reading the code. This `hiveConf` is specifically for Hive Sync, even it's not `HiveConf` type anymore, but `hiveConf` will still represent what it does? WDYT?

vinothchandar replied:
I still think its confusing to have it as hiveConf...  esp when you are actually creating ` HiveConf` from this down below..

garyli1019 replied:
changed to `conf`

pratyakshsharma commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/deltastreamer/HoodieDeltaStreamer.java:
```diff
@@ -90,35 +90,33 @@
-    this.cfg = cfg;
-    this.deltaSyncService = new DeltaSyncService(cfg, jssc, fs, hiveConf, properties);
+  public HoodieDeltaStreamer(Config cfg, JavaSparkContext jssc, FileSystem fs, Configuration hiveConf) throws IOException {
+    this(cfg, jssc, fs, hiveConf, null);
```
lets change the variable name from hiveConf to conf?

vinothchandar replied:
same here.

pratyakshsharma commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/deltastreamer/HoodieDeltaStreamer.java:
```diff
@@ -90,35 +90,33 @@
   }
 
-  public HoodieDeltaStreamer(Config cfg, JavaSparkContext jssc, FileSystem fs, HiveConf hiveConf) throws IOException {
+  public HoodieDeltaStreamer(Config cfg, JavaSparkContext jssc, FileSystem fs, Configuration hiveConf,
```
ditto.

pratyakshsharma commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/deltastreamer/HoodieDeltaStreamer.java:
```diff
@@ -371,7 +379,7 @@ public static void main(String[] args) throws Exception {
     private transient DeltaSync deltaSync;
 
-    public DeltaSyncService(Config cfg, JavaSparkContext jssc, FileSystem fs, HiveConf hiveConf,
+    public DeltaSyncService(Config cfg, JavaSparkContext jssc, FileSystem fs, Configuration hiveConf,
```
ditto.

vinothchandar commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/deltastreamer/HoodieDeltaStreamer.java:
```diff
@@ -90,35 +90,33 @@
   public HoodieDeltaStreamer(Config cfg, JavaSparkContext jssc) throws IOException {
     this(cfg, jssc, FSUtils.getFs(cfg.targetBasePath, jssc.hadoopConfiguration()),
-        getDefaultHiveConf(jssc.hadoopConfiguration()));
+        jssc.hadoopConfiguration(), null);
```
can we not pass `null` as a sentinel.. its risky business.. would an empty properties/map work?

garyli1019 replied:
totally understand

garyli1019 replied:
unfortunately, the `null` was served as a flag in many places in the delta streamer, because the command-line tool will produce `null`. If we want to change this, it might need a code refactoring.

vinothchandar replied:
got it..

vinothchandar commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/deltastreamer/HoodieDeltaStreamer.java:
```diff
@@ -90,35 +90,33 @@
-  public HoodieDeltaStreamer(Config cfg, JavaSparkContext jssc, FileSystem fs, HiveConf hiveConf) throws IOException {
+  public HoodieDeltaStreamer(Config cfg, JavaSparkContext jssc, FileSystem fs, Configuration hiveConf,
+                             TypedProperties properties) throws IOException {
+    if (cfg.initialCheckpointProvider != null && cfg.bootstrapFromPath != null && cfg.checkpoint == null) {
```
let's use Option instead of nulls?

garyli1019 replied:
same as above, it will need a code refactoring if we wanna get rid of null

vinothchandar replied:
lets file a separate issue for this? :)

garyli1019 replied:
yes https://issues.apache.org/jira/browse/HUDI-791

vinothchandar commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/deltastreamer/HoodieDeltaStreamer.java:
```diff
@@ -90,35 +90,33 @@
+    if (cfg.initialCheckpointProvider != null && cfg.bootstrapFromPath != null && cfg.checkpoint == null) {
+      InitialCheckPointProvider checkPointProvider =
+          UtilHelpers.createInitialCheckpointProvider(cfg.initialCheckpointProvider, new Path(cfg.bootstrapFromPath), fs);
+      cfg.checkpoint = checkPointProvider.getCheckpoint();
```
IIUC setting `cfg.checkpoint` will force use of that timestamp instead of what we normally do - read from the last commit? 

Should we do this only when creating the dataset for the first time..

garyli1019 replied:
I think this depends on how we design the migration flow for the user.
What I did myself is I use Spark datasource to do a bulkInsert to convert all the plain parquet files to Hudi format, then the second job I'd like to use delta streamer to read from Kafka. So this initialCheckpointProvider should be the first delta streamer job when switching sources.

vinothchandar replied:
yes.. do you think if we made it such that even if someone runs delta streamer few times after initial bootstrap, the initial checkpoint provider is used just once?  otherwise, you need to scramble to stop the delta streamer after the first run or manually run it by hand once before scheduling it using airflow or deploying in --continuous mode?

garyli1019 replied:
I believe the initial checkpoint provider should be just used once when the user wants to switch from one source to another. After that, the delta streamer should be able to get the checkpoint from the previous commit. We can improve this once the bootstrap is ready. At this point, I am not sure how to put everything together if we want one step to handling everything.

vinothchandar replied:
okay.. lets revisit once we have bootstrap support.. cc @bvaradar as fyi

vinothchandar commented on hudi-utilities/src/test/java/org/apache/hudi/utilities/TestHoodieDeltaStreamer.java:
```diff
@@ -394,6 +394,26 @@ public void testProps() {
   }
 
+  @Test
+  public void testInitialCheckpointProvider() throws IOException {
```
testKafkaConnectCheckpointProvider?

garyli1019 replied:
done

vinothchandar commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/deltastreamer/HoodieDeltaStreamer.java:
```diff
@@ -293,6 +295,12 @@ public Operation convert(String value) throws ParameterException {
     @Parameter(names = {"--checkpoint"}, description = "Resume Delta Streamer from this checkpoint.")
     public String checkpoint = null;
 
+    @Parameter(names = {"--bootstrap-from"}, description = "Initial bootstrap from this path")
```
this is a path where the current dataset resides? I can see that we can eventually use this for actually bootstrapping the dataset.. but wondering if for now, the CheckpointProvider just takes a property containing the base path for reading/.computing checkpoints? 

i.e we can remove `--bootstrap-from`.. it almost sounds like we are actually bootstrapping the data.

garyli1019 replied:
yea, with the `props` file, we can get the path there. Don't need this field anymore

vinothchandar commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/checkpointing/InitialCheckPointProvider.java:
```diff
@@ -20,12 +20,23 @@
+  protected final Path path;
+  protected final FileSystem fs;
+
+  public InitialCheckPointProvider(final Path basePath, final FileSystem fileSystem) {
```
it may be worth turning this into  `abstract void init(Path, FileSystem)` method? that way all subclasses are forced to implement the right one.. 

Also can we pass the `props` or the master list of properties (like we do for key generators) into this abstraction and let the CheckpointProvider use an explicit property like `hoodie.deltastreamer.checkpointprovider.kafka.connect.path` to derive the bootstrap path

garyli1019 replied:
Adding `props` is a good idea. It will add more flexibility for other providers.

garyli1019 commented:
hmm... Looks like the checkstyle auto-fix something...Let me see what's going...

garyli1019 commented:
I added the save action and checkstyle as documented, not sure which one triggered all those `final` and `this`. Is that ok to add those? As a scala programmer I do prefer to use `val` all the time and the Google Java Guide does encourage using final, but I am not sure what is the preference of Hudi.

garyli1019 commented:
Addressed some comments, summary:

- Removed `--bootstrap-from` option in the delta streamer. Use `hoodie.deltastreamer.checkpoint.provider.path` field in the props instead. 
- Use TypedProperty to construct the `InitialCheckPointProvider` and `init(FileSystem fs)` to initialize the class
- Keep `hiveConf` as the variable name even change `HiveConf` type to `Configuration`. Open to discussion if you guys don't agree.
- Not able to replace all `null` in this PR because `null` was served as a flag in the delta streamer workflow, this might change the behavior of other classes using the `TypedProperty` field. Will need a separate PR to do the code refactoring. 
- The style check tool automatically adds `final` and `this` to match the stylecheck.xml.

vinothchandar commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/UtilHelpers.java:
```diff
@@ -319,18 +331,18 @@ private static Boolean tableExists(Connection conn, Map<String, String> options)
-    String table = options.get(JDBCOptions.JDBC_TABLE_NAME());
-    boolean tableExists = tableExists(conn,options);
+  public static Schema getJDBCSchema(final Map<String, String> options) throws Exception {
+    final Connection conn = createConnectionFactory(options);
```
I get where you are coming from for `final` but given we are not following this everywhere.. can we please limit this PR to the minimal changes needed for this functionality..

vinothchandar commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/UtilHelpers.java:
```diff
@@ -219,13 +230,13 @@ public static JavaSparkContext buildSparkContext(String appName, String sparkMas
-   * @param jsc Java Spark Context
-   * @param basePath Base Path
-   * @param schemaStr Schema
+   * @param jsc         Java Spark Context
```
are these from intellij formatting...

garyli1019 replied:
yes, these might cause by the different settings when I switched machines. I removed all `this` and `final` manually instead of reverting commit, so there are still a few lines changing due to the initial formatting. I can revert those comments as well if it's preferable.

vinothchandar replied:
Thanks for doing it.. looks okay for now.

vinothchandar commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/UtilHelpers.java:
```diff
@@ -219,13 +230,13 @@ public static JavaSparkContext buildSparkContext(String appName, String sparkMas
    */
   public static HoodieWriteClient createHoodieClient(JavaSparkContext jsc, String basePath, String schemaStr,
-      int parallelism, Option<String> compactionStrategyClass, TypedProperties properties) {
+                                                     int parallelism, Option<String> compactionStrategyClass, TypedProperties properties) {
```
I think there is a setting not to format the lines untouched.. may be we can avoid all these whitespace changes that way..

vinothchandar commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/checkpointing/InitialCheckPointProvider.java:
```diff
@@ -18,14 +18,38 @@
+   *
+   * @param fileSystem
+   */
+  public abstract void init(FileSystem fileSystem);
```
would `Configuration` be a better choice here? is more general and the user can construct the filesystem from that? It will also carry other configurations picked up sparkContext say aws creds etc

garyli1019 replied:
Good point. now it's not `hiveConf` any more :)

vinothchandar commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/deltastreamer/DeltaSync.java:
```diff
@@ -130,7 +130,7 @@
    * Hive Config.
    */
-  private transient HiveConf hiveConf;
+  private transient Configuration conf;
```
:). thank you.

garyli1019 commented:
@vinothchandar Thanks for all the feedback! Very helpful! 
Comments addressed. Please take a look.

vinothchandar commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/deltastreamer/HoodieDeltaStreamer.java:
```diff
@@ -293,6 +296,12 @@ public Operation convert(String value) throws ParameterException {
+    @Parameter(names = {"--initial-checkpoint-provider"}, description = "Generate check point for delta streamer "
+        + "for the first run. This field will override the checkpoint of last commit using the checkpoint field. "
+        + "Use this field only when switch source, for example, from DFS source to Kafka Source. Check the class "
+        + "org.apache.hudi.utilities.checkpointing for details")
```
`InitialCheckPointProvider` did you intend to write the name of the class here?

garyli1019 replied:
Do you mean we should add `InitialCheckPointProvider` here or we should remove this description?

garyli1019 replied:
Changed the description to match with `--schemaprovider-class`
